src/url.rs

```diff
@@ -7,18 +7,74 @@ const QUOTES: &[char] = &['\'', '\"'];
 /// Scan for URLs starting from the trigger character ":", requires "://".
 ///
 /// Based on RFC 3986.
-pub struct UrlScanner {}
+pub struct UrlScanner;
 
 impl Scanner for UrlScanner {
-    fn scan(&self, s: &str, colon: usize) -> Option<Range<usize>> {
-        let after_slash_slash = colon + 3;
+    /// Scan for an URL at the given separator index in the string.
+    ///
+    /// The kind of separator that was used (`://` vs `.`) has effect on whether URLs with no
+    /// schemes are found.
+    ///
+    /// Returns `None` if none was found, or if an invalid separator index was given.
+    fn scan(&self, s: &str, separator: usize) -> Option<Range<usize>> {
+        // There must be something before separator for scheme or host
+        if separator == 0 {
+            return None;
+        }
+
+        // Detect used separator, being `://` or `.`
+        let (is_slash_slash, separator_len) = if s[separator..].starts_with("://") {
+            (true, "://".len())
+        } else if s[separator..].starts_with('.') {
+            (false, ".".len())
+        } else {
+            return None;
+        };
+        let after_separator = separator + separator_len;
+
+        // When scanning a dot, for URLs without a scheme, do some addadditional checks
+        if !is_slash_slash {
+            // The TLD must be at least 2 characters long
+            let end = &s[after_separator..];
+            let after_last_dot = end
+                .char_indices()
+                .take_while(|(_, c)| *c != '/')
+                .filter(|(_, c)| *c == '.')
+                .last()
+                .map(|(i, _)| i + 1)
+                .unwrap_or(0);
+            let tld_too_short = end[after_last_dot..]
+                .chars()
+                .take_while(|c| c.is_ascii_alphabetic())
+                .take(2)
+                .count()
+                < 2;
+            if tld_too_short {
+                return None;
+            }
+
+            // If this looks like an email address, don't scan it as URL
```

Comment:
Logic to detect whether an URL (without scheme) is an email address, in which case it won't be recognized as URL.

I'm not too happy about this myself, but don't see a clear way on how to improve this. Do you think this is good enough?

Reply:
Was trying to refactor the code a bit by putting it into a separate function:

```rust
#[inline]
fn is_mail(input: &str) -> bool {
    // Mail address must always contain an `@` sign. Return early if not.
    let at_pos = match input.chars().position(|c| c == '@') {
        Some(at_pos) => at_pos,
        None => return false,
    };
    // TODO: Why is this a special case?
    // Shouldn't colons be allowed in a mail address?
    if input[0..at_pos].contains(':') {
        return false;
    }
    let domain = &input[at_pos + 1..];
    if let Some(dot_pos) = domain.chars().position(|c| c == '.') {
        if self.scan(domain, dot_pos).is_some() {
            return true;
        }
    }
    false
}
```

The idea was to return early to keep the control flow on the left.
Also, maybe we could add some comments to help newcomers understand the logic.
There is still room for improvement I guess. 🤷‍♀️

Reply:
The main advantage would be that we could write unit tests for the mail detection. 😉

Reply:
Good idea to extract it. The `scan` function was getting large.

> // TODO: Why is this a special case?
> // Shouldn't colons be allowed in a mail address?

An URL may contain user credentials in the form of `user:pass@host/`, while an email address cannot. This case might be so uncommon though that we can skip it all together.

Reply:
> Shouldn't colons be allowed in an email address?

It seems you're right. According to RFC 532{1,2}

> space and `"(),:;<>@[\]` characters are allowed with restrictions (they are only allowed inside a quoted string, as described in the paragraph below, and in addition, a backslash or double-quote must be preceded by a backslash);

It notes the `:` is only allowed within a quoted string, along with the `@`. This makes handling it quite a bit more complex. I'll just remove the special `:` case, which should make it 'good enough'.

The funny thing is, this library already includes an email address scanner. I'll try to see if I can easily integrate it here first for a simple `is_mail` check without reinventing the wheel.

Reply:
I've added `is_mail` as helper to the email scanner module:

```rust
/// Helper function to check if given string is considered an email address.
#[inline]
pub(crate) fn is_mail(input: &str) -> bool {
    input
        .char_indices()
        .filter(|(_, c)| *c == '@')
        .any(|(i, _)| {
            let scanner = EmailScanner {
                domain_must_have_dot: true,
            };
            scanner.scan(input, i).is_some()
        })
}
```

Now I simply do:

```rust
// If this is an email address, don't scan it as URL
if email::is_mail(&s[after_separator..]) {
    return None;
}
```

This is a lot cleaner.